packages/components/src/TreeView/TreeViewItem/TreeViewItem.component.js

```diff
@@ -136,6 +136,13 @@ class TreeViewItem extends React.Component {
 
 		const paddingLeft = `${depth * PADDING_NORMAL + PADDING_LARGE}px`;
 
+		let iconProps = !icon || typeof icon.toString() === icon ? { name: getItemIcon(icon, toggled) } : icon;
```

Comment:
Wow!

`const` instead of `let` here but...
I saw what you want after reading it 5 times. 
```jsx
let iconProps = {};
if((typeof icon === 'object') && (icon !== null)) {
  // it's your icon object
  iconProps = Object.assign({}, icon); // immutability concern
} else {
  // simple text, for compatibility
  iconProps.name = getItemIcon(icon, toggled);
}
iconProps.className = classnames(css['tc-treeview-folder'], icon.className);
```

Reply:
updated to "const" and avoiding the error of accessing icon in this snippet